web/impact/impact/v1/views/user_program_confirmation_detail_view.py

```diff
@@ -0,0 +1,101 @@
+from django.contrib.auth import get_user_model
+from django.template import loader
+from rest_framework.exceptions import ParseError
+from rest_framework.response import Response
+
+from accelerator.models import (
+    UserRole,
+    Program,
+    ProgramRoleGrant,
+    ProgramRole
+)
+from impact.minimal_email_handler import MinimalEmailHandler
+from impact.v1.helpers import UserProgramConfirmationHelper
+from impact.v1.views.base_detail_view import BaseDetailView
+from impact.v1.views.user_detail_view import NO_USER_ERROR
+
+from impact.permissions.v1_api_permissions import UserDetailViewPermission
+
+User = get_user_model()
+
+SUBJECT = '[MassChallenge] Confirmed: Your Mentorship Commitment'
+MENTOR_SHIP_COMMITMENT_EMAIL_TEMPLATE = 'emails/mentorship_commitment_email.html'
+INVALID_INPUT_ERROR = "'{}' must be a list of numbers."
+
+
+def validate_input(request, key):
+    try:
+        return list(map(int, request.data.get(key, [])))
+    except Exception:
+        raise ParseError(INVALID_INPUT_ERROR.format(key))
+
+
+class UserProgramConfirmationDetailView(BaseDetailView):
+    view_name = "user_program_confirmation_detail"
+    helper_class = UserProgramConfirmationHelper
+    permission_classes = (UserDetailViewPermission,)
+
+    def __init__(self, *args, **kwargs):
+        self.user = None
+        super().__init__(*args, **kwargs)
+
+    def add_programrolegrant(self, program_ids=None, user_role=None):
+        programs = Program.objects.filter(pk__in=program_ids)
+        user_role_obj = UserRole.objects.get(name=user_role)
+
+        programrolegrants = [
+            ProgramRoleGrant(
+                person=self.user,
+                program_role=ProgramRole.objects.filter(
+                    program=program,
+                    user_role=user_role_obj
+                ).first())
+            for program in programs
+        ]
+        ProgramRoleGrant.objects.bulk_create(
```

Comment:
I appreciate that you thought to use bulk_create to spare the DB, but you're still hitting the DB in the loop. We could change that by doing something like:
```
program_roles = dict(ProgramRole.objects.filter(user_role=user_role).order_by('-id').values_list('program_id', 'id')
```
before the loop starts and then specify 
`program_role_id=program_roles[program_id]` 
as we create each ProgramRoleGrant in the loop. 

This would also allow you to loop over `program_ids`, rather than getting the programs from the db. 

(the `order_by` is a bit of a hack - in case staff create later ProgramRoles with the UserRoles we're interested in , this should ensure that the dict has the first created ProgramRole with this UserRole, which will be the one we want)

Also, as long as I'm nerding out on db access, I notice that you use a throwaway `get` to retrieve the user_role_obj by name rather than using `user_role__name=user_role` in the ProgramRoleGrant creation. Is this to avoid the join?

Reply:
didn't think of this. great suggestion